common/src/main/java/com/microsoft/hydralab/common/management/DeviceManager.java

```diff
@@ -254,7 +253,13 @@ public void updateAllDeviceInfo() {
     abstract public void quitMobileAppiumDriver(DeviceInfo deviceInfo, Logger logger);
 
     abstract public void execCommandOnDevice(DeviceInfo deviceInfo, String command, Logger logger);
-    
+
+    public void execCommandOnAgent(DeviceInfo deviceInfo, String command, Logger logger) {
+        ITestRun testRun = TestRunThreadContext.getTestRun();

```

Comment:
please assert the value of testRun and getResultFolder. Otherwise, the parseHydraLabVariable method maybe NPE

Assert.notNull(testRun, "There is no testRun instance in ThreadContext!");
        Assert.notNull(testRun.getResultFolder(), "The testRun instance in ThreadContext does not have resultFolder property!");

Reply:
Or just verify in the parseHydraLabVariable method. execCommandOnAgent may be called when setup agent or other time

Reply:
Revised, Please help review